lib/study_cleanup_tools.rb

```diff
@@ -0,0 +1,172 @@
+  #   - +allow_dev_env+: (Boolean) => allow execution of this method in development environment (default: false)
+  #
+  # * *raises*
+  #   - (RuntimeError) => if command cannot be run for failing a security check (hostname, project, environment, etc)
+  #
+  # * *returns*
+  #   - (Integer) => count of workspaces that were deleted
+  def self.delete_all_orphaned_workspaces(project_name = FireCloudClient::PORTAL_NAMESPACE, allow_dev_env: false)
+    raise_exception_unless_true(:permit_environment?) { permit_environment?(allow_dev_env) }
+    raise_exception_unless_true(:permit_hostname?) { permit_hostname? }
+    raise_exception_unless_true(:permit_billing_project?) { permit_billing_project?(project_name) }
+    raise_exception_unless_true(:is_continuous_integration?) { is_continuous_integration? }
+    raise_exception_unless_true(:confirm_delete_request?) { confirm_delete_request? } if !!(allow_dev_env && Rails.env.development?)
+    workspaces = ApplicationController.firecloud_client.workspaces(project_name)
+    workspace_count = 0
+    workspaces.each do |workspace|
+      ws_attr = workspace.dig('workspace')
+      ws_name = ws_attr['name']
+      ws_project = ws_attr['namespace']
+      ws_owner = ws_attr['createdBy']
+      existing_study = Study.find_by(firecloud_project: ws_project, firecloud_workspace: ws_name)
+      if existing_study.present?
+        puts "skipping #{ws_project}/#{ws_name} as it belongs to #{existing_study.accession}"
+        next
+      else
+        begin
+          if service_account_created_workspace?(workspace)
+            print "deleting #{ws_project}/#{ws_name}... "
+            ApplicationController.firecloud_client.delete_workspace(ws_project, ws_name)
+            workspace_count += 1
+            puts "#{ws_project}/#{ws_name} successfully deleted"
+          else
+            puts "skipping #{ws_project}/#{ws_name} because it was created by #{ws_owner}"
+          end
+        rescue => e
+          puts "Unable to delete #{ws_project}/#{ws_name} due to: #{e.class.name} - #{e.message}"
+        end
+      end
+    end
+    workspace_count
+  end
+
+  ##
+  # security safeguards to prevent accidental deletes of production/staging workspaces and studies
+  ##
+
+  # handler to run a security check and raise an error on returning false to halt execution
+  #
+  # * *params*
+  #   - +validation_name+: (String) => name of validation being run (for error message)
+  #   - +&block+: (Proc) => validation method to execute that returns true/false
+  #
+  # * *raises*
+  #   - (RuntimeError) => raise exception if security check returns false to halt parent method execution
+  def self.raise_exception_unless_true(validation_name, &block)
+    valid_command = yield
+    if !valid_command
+      # report validation that failed and method caller
+      error_message = "halting execution as #{validation_name} failed validation when calling #{self.name}##{caller_locations.first.label}"
```

Comment:
this is firmly in the realm of 'nit'.  But you're not actually halting execution here--you're just throwing the exception. (which might be caught further upstream).   So I'd remove 'halting execution' from the error_message.

Reply:
Fair enough.